File: server/php/src/Presentation/Validators/Payment/PaymentValidator.php
<?php

namespace OnlinePayments\ExampleApp\Presentation\Validators\Payment;

use OnlinePayments\ExampleApp\Application\Domain\Enums\Common\Currency;
use OnlinePayments\ExampleApp\Application\Domain\Enums\Payments\PaymentMethodType;
use OnlinePayments\ExampleApp\Configuration\Exceptions\ValidationException;
use OnlinePayments\ExampleApp\Presentation\Models\Payment\Request;
use OnlinePayments\ExampleApp\Presentation\Validators\Rules\CardRules;
use OnlinePayments\ExampleApp\Presentation\Validators\Rules\CountryRules;
use OnlinePayments\ExampleApp\Presentation\Validators\Rules\MandateRules;

final class PaymentValidator
{
    /**
     * @throws ValidationException
     */
    public function validate(Request $request): array
    {
        $errors = [];

        if (!empty($request->hostedTokenizationId)) {
            if ($request->card !== null) {
                $cardEmpty = (
                    empty($request->card->number)
                    && empty($request->card->holderName)
                    && empty($request->card->verificationCode)
                    && empty($request->card->expiryMonth)
                    && empty($request->card->expiryYear)
                );

                if (!$cardEmpty) {
                    $errors[] = 'If hosted tokenization id is provided, card details must not be filled.';
                }
            }
        }

        if (!isset($request->amount)) {
            $errors[] = 'The Amount field is required.';
        }
        elseif (!is_numeric($request->amount)) {
            $errors[] = 'The Amount field must be a number.';
        }
        elseif ((int)$request->amount <= 0) {
            $errors[] = 'The Amount field must be greater than zero.';
        }

        if (!$request->currency) {
            $errors[] = 'The Currency field is required.';
        }
        elseif (Currency::tryFrom($request->currency) === null) {
            $errors[] = 'The Currency field is invalid.';
        }

        $methodEnum = isset($request->method) ? PaymentMethodType::tryFrom($request->method) : null;
        if (!$request->method) {
            $errors[] = 'The Method field is required.';
        }
        elseif ($methodEnum === null) {
            $errors[] = 'The Method field is invalid.';
        }

        if ($methodEnum === PaymentMethodType::CREDIT_CARD) {
            $errors = array_merge($errors, CardRules::validateCard($request->card));

            if(isset($request->billingAddress->country)){
                $errors = array_merge($errors, CountryRules::validateCountry($request->billingAddress->country, 'BillingAddress'));
            }

            if(isset($request->billingAddress->country)){
                $errors = array_merge($errors, CountryRules::validateCountry($request->billingAddress->country, 'BillingAddress'));
            }
        }

        if ($methodEnum === PaymentMethodType::DIRECT_DEBIT) {
            $errors = array_merge($errors, MandateRules::validateMandate($request->mandate));
        }

        if (!empty($errors)) {
            throw new ValidationException($errors);
        }

        return $errors;
    }
}

File: server/python/app/presentation/validators/payment/payment_validator.py
from app.application.domain.enums.common.currency import Currency
from app.application.domain.enums.payments.payment_method_type import PaymentMethodType
from app.application.exceptions.validation_exception import ValidationException
from app.presentation.models.create_payment.request import CreatePaymentRequest
from app.presentation.validators.common.address.address_validator import AddressValidator
from app.presentation.validators.common.card.card_validator import CardValidator
from app.presentation.validators.common.mandate.mandate_validator import MandateValidator

class PaymentValidator:

    def validate(self, request: CreatePaymentRequest) -> None:
        errors = []

        if request.amount is None:
            errors.append("The Amount field is required.")
        else:
            try:
                amount = float(request.amount)
                if amount <= 0:
                    errors.append("The Amount field must be greater than zero.")
                elif round(amount * 100) != amount * 100:
                    errors.append("The Amount field is invalid.")
            except (ValueError, TypeError):
                errors.append("The Amount field must be a number.")

        if not request.currency:
            errors.append("The Currency field is required.")
        elif not Currency.has_value(request.currency):
            errors.append("The Currency field is invalid.")

        if not request.method:
            errors.append("The Method field is required.")
        elif not PaymentMethodType.has_value(request.method):
            errors.append("The Method field is invalid.")
        else:
            if request.method == PaymentMethodType.TOKEN.value:
                if not request.hosted_tokenization_id:
                    errors.append("The HostedTokenizationId field is required when payment method is TOKEN.")

            if request.hosted_tokenization_id and request.card:
                card = request.card
                card_empty = (
                    not (card.number or "").strip() and
                    not (card.holder_name or "").strip() and
                    not (card.verification_code or "").strip() and
                    not str(card.expiry_month or "").strip() and
                    not str(card.expiry_year or "").strip()
                )
                if not card_empty:
                    errors.append("If hosted tokenization id is provided, card details must not be filled.")

            if request.method == PaymentMethodType.CREDIT_CARD.value:
                errors.extend(CardValidator().validate(request.card))

            if request.method == PaymentMethodType.DIRECT_DEBIT.value:
                if request.mandate is None:
                    errors.append("The mandate object is required.")
                else:
                    errors.extend(MandateValidator().validate(request.mandate))

        if request.billing_address and any(vars(request.billing_address).values()):
            errors.extend(AddressValidator.validate(request.billing_address, "BillingAddress"))

        if request.shipping_address and any(vars(request.shipping_address).values()):
            errors.extend(AddressValidator.validate(request.shipping_address, "ShippingAddress"))

        if errors:
            raise ValidationException(errors)